keps/sig-node/1592-pod-level-resource-limits/README.md

```diff
@@ -0,0 +1,439 @@
+This proposal is compatible with both cgroup v1 and cgroup v2. Both versions of cgroup allow specifying limits on all levels of the cgroup hierarchy.
+
+#### ResourceOverhead
+
+Since the runtime shim is started in the pod-level cgroup, there should be no bad interactions between this proposed feature and the ResourceOverhead feature.
+
+The `ResourceOverhead` feature can be leveraged to allow cluster administrators to automatically increase the amount of resources allowed for a pod. This enables creating cluster-wide policies that inject sidecars to a pod without risking that the pod becomes starved for resources due to the pod-level resource limit.
+
+#### Memory-based Emptydir volumes
+
+This proposal doesn't change the current status quo in any way. Memory used for files in a `tmpfs` volume is accounted for as shared-memory. The Linux kernel charges the used memory to cgroup that first touched the shared page for the memory. Moving the charge for shared memory pages between cgroups is not currently supported by the Linux kernel. Defining a memory limit on the pod cgroup level would not change the underlying limitation that memory used by files in the volume are accounted to the cgroup that first touched the memory.
+
+#### HugeTLB cgroup
+
+The hugetlb cgroup is similar to the `memory` cgroup, except that in contrast to memory where unset means unlimited, when `hugetlb` is unset it is essentially the same as enforcing 0. 
+
+The behavior suggested by this KEP then becomes:
+
+1. If no pod-level limits are specified for `hugetlb`, make no change to the current implementation.
+1. If pod-level limits are set, apply them at the pod-level cgroup.
+    1. for each container in the pod, 
+         1. if container-level limits are set, use the values as provided
+         1. if container-level limits are NOT set, then set the container-level cgroup to unlimited.
+
+As on the container level, the pod level request must be equal to the pod level limit to be considered valid.
+
+### PID, and other cgroups
+
+These are out of scope for this proposal. This proposal doesn't remove the ability to set these limits and requests on the container level, if this was already supported by Kubernetes before.
+
+#### Init Containers
+
+The current implementation (before this proposal) sets up the pod-level cgroup using the maximum between the sum of all the sidecars or the biggest limit for any single Init container. 
+
+There are a few options: 
+1. Limit the pod-level cgroup only after the `InitContainers` have successfully terminated
+1. Require the pod-level cgroup (if it exists) to apply to the `InitContainers` as well.
+1. Allow specifying pod-level restrictions separately for Init containers and sidecars.
+
+Since init containers run sequentially, there is no requirement to share resources with any other container. Therefore the first option is the best - the pod-level cgroup should be restricted only after all of the init containers have finished running.
```

Comment:
I disagree. I think we should go with the second option for consistency (apply pod-level cgroup to init containers). Consider the case where pod-level limits are defined, and the user didn't bother to set container-level limits. I would expect the pod limits to cap the init containers.

Reply:
@tallclair I agree with you from the UX point of view. 

One problem I see is that if the init container requires much more resources than the regular + sidecar containers, users will be forced to set a very high pod resource limits to accommodate initial surge. This might not be a concern given that we [already account for](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#resource-sharing-within-containers) the init container resource requirements during scheduling, but just want to bring it up as it's legit use case.